### PR TITLE
feature(flowchart): support `linkStyle -` to style the previous line defined

### LINKED ISSUE #431: Styling specific lines
Is it possible to style specific lines? I can see there is general class for lines but it will apply to all. I'm trying to highlight specific path in the flow chart.

Thanks,
Radmila

### PR DESCRIPTION
**Pull request by Zumbala**
## :bookmark_tabs: Summary

`linkStyle` can now be used without a number, use a minus instead to set the style of the last line defined.

Resolves #431

## :straight_ruler: Design Decisions

Changed the jison to alllow minus as a replacement for the number

    | LINKSTYLE SPACE MINUS SPACE stylesOpt
          {$$ = $1;yy.updateLink([yy.getEdges().length-1],$5);}

sample:

      ---
      title: Traffic light
      ---
      flowchart TD
          Driving --> TrafficLight{ Orange? }
          TrafficLight --> |No| Go{Green?}
          linkStyle - stroke:blue;
          TrafficLight ---> |Yes| SpeedUp
          linkStyle - stroke:orange;
          Go --> |Yes| JustCruise
          linkStyle - stroke:green;
          Go --> |No| Breaks[Hit the brakes]
          linkStyle - stroke:red;

Graph-66-should-render-a-flowchart-with-linkStyle---no-htmlLabels snap

Make sure you

- [ ] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added unit/e2e tests (if appropriate)
- [x] :notebook: have added documentation (if appropriate)
- [x] :bookmark: targeted `develop` branch

### CODE REVIEW

**Comment by Zumbala** [comment]
@sidharthv96 I did what you asked of me to sync, but I still get an error, which in my opinion is unrelated to my changes.

I am not that well educated in GIT, so maybe I am doing some things incorrect.

Hope you can help me again in solving below issue?

        ⎯⎯⎯⎯⎯⎯⎯ Failed Tests 1 ⎯⎯⎯⎯⎯⎯⎯
       
        FAIL  packages/mermaid/src/diagram.spec.ts > diagram detection > should throw the right error for incorrect diagram
        Error: Snapshot `diagram detection > should throw the right error for incorrect diagram 1` mismatched


          - Expected  - 1
          + Received  + 1

             `"Parse error on line 2:
             graph TD; A-->
            --------------^
          - Expecting 'AMP', 'ALPHA', 'COLON', 'PIPE', 'TESTSTR', 'DOWN', 'DEFAULT', 'NUM', 'COMMA', 'MINUS', 'BRKT', 'DOT', 
          - 'PUNCTUATION', 'UNICODE_TEXT', 'PLUS', 'EQUALS', 'MULT', 'UNDERSCORE', got 'EOF'"`
          + Expecting 'AMP', 'ALPHA', 'COLON', 'PIPE', 'TESTSTR', 'DOWN', 'DEFAULT', 'MINUS', 'NUM', 'COMMA', 'BRKT', 'DOT', 
         'PUNCTUATION', 'UNICODE_TEXT', 'PLUS', 'EQUALS', 'MULT', 'UNDERSCORE', got 'EOF'"`
         ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯

File: packages/mermaid/src/diagrams/flowchart/parser/flow.jison [hunk 1]
```diff
@@ -494,6 +494,8 @@ styleStatement:STYLE SPACE alphaNum SPACE stylesOpt
 linkStyleStatement
     : LINKSTYLE SPACE DEFAULT SPACE stylesOpt
           {$$ = $1;yy.updateLink([$3],$5);}
+    | LINKSTYLE SPACE MINUS SPACE stylesOpt
+          {$$ = $1;yy.updateLink([yy.getEdges().length-1],$5);}
```
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/flowchart/parser/flow.jison:498]: Is the `-` necessary? 
Can we simply write 

```
A --> B
linkStyle stroke:red;
B --> C
linkStyle stroke:green;
```

**Comment by rockmasterflex69** [comment]
What's the status on this? It would be super great to have style be able to be programmatically linked to an edge without having to memorize the indices and hardcode them, and this is certainly a good step in that direction - allowing people to define the style after the edge.

File: packages/mermaid/src/diagrams/flowchart/parser/flow.jison [hunk 2]
```diff
@@ -540,16 +540,22 @@ styleStatement:STYLE SPACE idString SPACE stylesOpt
 linkStyleStatement
     : LINKSTYLE SPACE DEFAULT SPACE stylesOpt
           {$$ = $LINKSTYLE;yy.updateLink([$DEFAULT],$stylesOpt);}
+    | LINKSTYLE SPACE MINUS SPACE stylesOpt
+          {$$ = $LINKSTYLE;yy.updateLink([yy.getEdges().length-1],$stylesOpt);}
```
Comment (nirname) [inline packages/mermaid/src/diagrams/flowchart/parser/flow.jison:544]: would not be better to wrap
```
edges.length - 1
```
in a function, e.g. `lastEdgeIndex` and put it in DB, instead of making some calculations here?

Multiple occurrences of the same code

File: packages/mermaid/src/diagrams/flowchart/parser/flow-lines.spec.js [hunk 3]
```diff
@@ -13,75 +13,134 @@ describe('[Lines] when parsing', () => {
...
 
     expect(edges[0].interpolate).toBe('basis');
     expect(edges[1].interpolate).toBe('cardinal');
   });
 
+  it('should handle line interpolation non-numbered with style', function () {
+    flow.parser.parse(
+      'graph TD\n' +
+        'A-->B\n' +
+        'linkStyle - interpolate basis stroke-width:1px;\n' +
```
Comment (nirname) [inline packages/mermaid/src/diagrams/flowchart/parser/flow-lines.spec.js:91]: Can we apply it twice?
```
linkStyle -
linkStyle -
```
I have not checked whether it should work or not, I am not well aware of flowchart behavior. Could you check it for me?

Comment (sidharthv96) [reply packages/mermaid/src/diagrams/flowchart/parser/flow-lines.spec.js:91]: Now you can actually play around with PR Version of mermaid in the deploy previews. 

It was actually added when I had similar questions on a PR and it was easier to test out all the edge cases live instead of either asking the author or checking out and building it locally.

Comment (nirname) [reply packages/mermaid/src/diagrams/flowchart/parser/flow-lines.spec.js:91]: The answer is yes, we can apply linkStyle twice and it overrides the previous one. It does not merge attributes. Resolving this, because linkStyle with index has the same behavior.

**Comment by nirname** [comment]
@rockmasterflex69 we are waiting for a reply from the author of PR. If it is abandoned than at some point one of the team members will probably fix it and. But I'd better not interfere, because there may be different circumstances and different amount of time on hands. So lets wait for the author a little bit. And if nothing happens we will decide what to do next. The original issue is open for everyone, so nothing, even the existence of other PR, prevents you from creating your own. You can use this as an example as well

File: packages/mermaid/src/diagrams/flowchart/parser/flow.jison (hunk 1 above)
Comment (sidharthv96) [reply packages/mermaid/src/diagrams/flowchart/parser/flow.jison]: Seems like it is needed to avoid some complications in the grammar.

**Comment by sidharthv96** [comment]
@nirname I think this is good to go now. Let's merge?

File: packages/mermaid/src/docs/syntax/flowchart.md [hunk 4]
```diff
@@ -643,8 +651,10 @@ Available curve styles include `basis`, `bumpX`, `bumpY`, `cardinal`, `catmullRo
 In this example, a left-to-right graph uses the `stepBefore` curve style:
 
 ```
+
 %%{ init: { 'flowchart': { 'curve': 'stepBefore' } } }%%
 graph LR
+
```
Comment (aloisklink) [inline packages/mermaid/src/docs/syntax/flowchart.md:657]: Hmmmm, I'm not sure what happened here to add these random newlines? Maybe a bad merge or some weird auto-formatting?

```suggestion
%%{ init: { 'flowchart': { 'curve': 'stepBefore' } } }%%
graph LR
```

**Review by aloisklink — APPROVED** [review]
Looks good to me!

By the way @Zumbala, your commits were made with the email: `<ha***en@ti**ix.com>` (I've hidden some characters, but you can find them by going to https://github.com/mermaid-js/mermaid/commit/125734228e7f163a38eb8b0874c857d052abfd6b.patch or in the `git log`).

This is completely optional, but if you want to link your contributions to your GitHub account and get a **Contributor** badge on this repo, you'll need to add your email, see [Adding an email address to your GitHub account](https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-email-preferences/adding-an-email-address-to-your-github-account).

**Review by nirname — CHANGES_REQUESTED** [review]
We have a breaking change. So we have to decide either to sacrifice one minor case, or fix it

**Comment by nirname** [comment]
@sidharthv96 we need to investigate it further. I'am not even sure that the feature that broke had been documented.

File: packages/mermaid/src/diagrams/flowchart/parser/flow.jison [hunk 5]
```diff
@@ -169,8 +169,8 @@ that id.
 "*"                   return 'MULT';
 "#"                   return 'BRKT';
 "&"                   return 'AMP';
+"-"                   return 'MINUS';
```
[the remark this answers is not in the document]
Comment (sidharthv96) [reply packages/mermaid/src/diagrams/flowchart/parser/flow.jison:172]: @nirname, isn't it possible to edit the grammar to support both usecase?

[the remark this answers is not in the document]
Comment (knsv) [reply packages/mermaid/src/diagrams/flowchart/parser/flow.jison:172]: The feature idea is brilliant! Cudos to @Zumbala  but this is a bloicker. 

Any existing diagrams with a minus in the label without quotes will suddenly break. I think there can be many of those. A minus sign is fairly common.
😞 

Let's try to modify the grammar to accommodate this!!! The idea is brilliant, and the implementation is good (apart from the blocker). 💯

**Comment by sidharthv96** [comment]
Converting to draft to avoid accidentally merging. 
Blocker : https://github.com/mermaid-js/mermaid/pull/4483#discussion_r1507434796

**Comment by nirname** [comment]
@Zumbala gentle reminder, any updates?

**Comment by voronoipotato** [comment]
@sidharthv96 I think it's safe to say that @Zumbala  is no longer active. This would be huge for making flowcharts and other graphs more useable. I wouldn't mind it even if all we could do is just have each edge be randomly colored and edges that have the same source node have the same color. @nirname  would you say this is unsalvageable?

**Comment by ramwin** [comment]
I love this pull request. Can any one merge this pull request?
